replications/internal/queue_management.go

```diff
@@ -400,17 +404,32 @@ func (qm *durableQueueManager) EnqueueData(replicationID platform.ID, data []byt
 	return nil
 }
 
-func (qm *durableQueueManager) newReplicationQueue(id platform.ID, queue *durablequeue.Queue) *replicationQueue {
+func (qm *durableQueueManager) newReplicationQueue(id platform.ID, orgID platform.ID, localBucketID platform.ID, queue *durablequeue.Queue) *replicationQueue {
 	logger := qm.logger.With(zap.String("replication_id", id.String()))
 	done := make(chan struct{})
 
 	return &replicationQueue{
-		id:           id,
-		queue:        queue,
-		done:         done,
-		receive:      make(chan struct{}, 1),
-		logger:       logger,
-		metrics:      qm.metrics,
-		remoteWriter: remotewrite.NewWriter(id, qm.configStore, qm.metrics, logger, done),
+		id:            id,
+		orgID:         orgID,
+		localBucketID: localBucketID,
+		queue:         queue,
+		done:          done,
+		receive:       make(chan struct{}, 1),
+		logger:        logger,
+		metrics:       qm.metrics,
+		remoteWriter:  remotewrite.NewWriter(id, qm.configStore, qm.metrics, logger, done),
+	}
+}
+
+// GetReplications returns the ids of all currently registered replication streams matching the provided orgID
+// and localBucketID
+func (qm *durableQueueManager) GetReplications(orgID platform.ID, localBucketID platform.ID) []platform.ID {
+	replications := make([]platform.ID, len(qm.replicationQueues))
```

Comment:
I believe initializing the slice like this will be problematic, and we should instead do `make([]platform.ID, 0, len(qm.replicationQueues))` for the filter operation.

Setting the length (second argument to `make`) will populate the slice with that many zero-value `platform.ID`. Appends will then add on to the slice. So in a case where only a few replications would match out of `qm.replicationQueues`, the returned slice will have a bunch of empty IDs, followed by the few matches. And if none were supposed to match, there'd be a non-zero-length slice returned that wouldn't work with the write function logic.

Using `make([]platform.ID, 0, len(qm.replicationQueues))` instead will set a slice with a 0 length but an underlying array large enough to possibly hold the full list without re-sizing, which should offer optimal performance for doing a filter like this.

Considering all this...we should also add a unit test for this function. That will flush out any problems with the slice initialization.

Aside: You would usually use slice initialization w/ setting the length like this if you knew exactly how long the result slice needed to be, and then you would loop over it by index and update its entries via the index. That would be for a transformation type of operation rather than a filter.

Reply:
Fixed the slice init and added a unit test for `GetReplications`. As for refactoring the `tracked` package - I like the first option, but agree on the complexity and time to refactor. The second option (I think) wouldn't eliminate the need for a struct because we still need to track the `maxQueueSizeBytes` as well, so two pieces of information with each replication id, rather than three. This is why I went with the third option of moving the `TrackedReplication` struct to the `replication.go` file in the `influxdb` package.